src/schoology-extractor/schoology_extractor/main.py

```diff
@@ -27,115 +27,72 @@
-    logger.info("Exporting sections - Getting courses")
-    courses_response = request_client.get_courses(page_size)
-    courses_list: List[Any] = []
-    while True:
-        courses_list = courses_list + courses_response.current_page_items
-        if courses_response.get_next_page() is None:
-            break
-
-    # now we can get a list of sections
-    course_ids = map(lambda x: x["id"], courses_list)
-    sections_list = request_client.get_section_by_course_ids(list(course_ids))
-    export_data.to_csv(sections_list, os.path.join(schoology_output_path, "sections.csv"))
-except Exception as ex:
-    logger.error('An exception has occurred in the process of generating the sections.csv file: %s', ex)
-
-
-# Export assigments
-logger.info("Exporting assigments")
-assignments = []
-try:
-    section_ids = map(lambda x: x["id"], sections_list)
-
-    assignments = request_client.get_assignments_by_section_ids(list(section_ids))
-
-    filtered_assignments = [
-        assignment
-        for assignment in assignments
-        if assignment["grading_period"] in grading_periods_array
-    ]
-
-    export_data.to_csv(
-        filtered_assignments, os.path.join(schoology_output_path, "assignments.csv")
-    )
-except Exception as ex:
-    logger.error('An exception has occurred in the process of generating the assigments.csv file: %s', ex)
-
-
-# Export submissions
-logger.info("Exporting submissions")
-try:
-    submissions_list: List[Any] = []
-
-    for assignment in assignments:
-        submissions_response = (
-            request_client.get_submissions_by_section_id_and_grade_item_id(
-                assignment["section_id"],
-                str(assignment["grade_item_id"]),
-                page_size
-            )
+def _create_file_from_list(action: Callable, file_name):
+    logger.info(f"Exporting {file_name}")
+    try:
+        data = action()
+        export_data.to_csv(data, os.path.join(schoology_output_path, file_name))
+    except Exception as ex:
+        logger.error(
+            f"An exception occurred while generating {file_name} : %s",
+            ex,
         )
-        while True:
-            submissions_list = submissions_list + submissions_response.current_page_items
-            if submissions_response.get_next_page() is None:
-                break
-
-    export_data.to_csv(
-        submissions_list, os.path.join(schoology_output_path, "submissions.csv")
-    )
-except Exception as ex:
-    logger.error('An exception has occurred in the process of generating the submissions.csv file: %s', ex)
+
+
+def _export_sections():
+    sections = facade.get_sections()
+    closure["sections"] = sections
+    return sections
+
+
+def _export_assignments():
+    assignments = facade.get_assignments(closure["sections"], grading_periods)
+    closure["assignments"] = assignments
+    return assignments
+
+
+def _export_submissions():
+    return facade.get_submissions(closure["assignments"])
+
+
+def main():
```

Comment:
Instead of using a closure object, I'd create just pure functions that return values, storing the results in simple variables and passing them as parameters where needed.
My only reason for this change, would be to keep things as declarative as possible.
Example:
Method `_export_sections()` is not exporting sections, it is returning a list of sections from the API and setting a value in the closure object. In case you wanted to keep that closure, I'd rename the function to something similar to `_get_sections_and_set_values_in_closure()`, keeping in mind that the function is actually doing two things: setting a value in the object, and returning a value that it gets from the API.

Reply:
I agree about renaming the methods - they're not doing the exporting.  I suppose I could simplify - instead of passing a Callable around, I could make the call and pass the data to the `_create_file...` functions. But I still have the problem of having the try/catch only once. I will think about it.

Reply:
Basically I'm following the strategy pattern by injecting the callable into the shared code that performs the actual export and wraps both the get and save with try/catch. But I needed to be able to pass some data around. The "closure" variable is scoped to the entire script and is thus immutable inside the function. While you can't set the value of that outer variable, you _can_ add entries to it. So `closure` is kind of a global container.

Reply:
@bradbanister would love to have your thoughts on this as well.

Reply:
For the closure object, it's similar to my "Result" object.  I'd give the closure object an actual type (I used a simple namedtuple) so mypy can validate fields.  I'd also rename to something more expressive...like "Result" 😉, or maybe ApiResults, etc.

For the facade, it looks like the primary purpose is around paging.  I'd again rename to something more expressive, maybe PagingDecorator?  I also wonder if it would be nicer as an actual Python decorator, meaning convert most of it to a function wrapper that does the paging and exception handling as a cross-cutting concern.

Reply:
Will rename `closure` and add types.

The `Facade` class's purpose is more than just paging. Name was definitely lazy and I'll work on a better one. Probably won't turn try to turn it into a decorator at this point, since it is performing business logic, not just cross-cutting concerns.

Reply:
Done

Reply:
@stephenfuqua  So, if going decorator is too much for now, you could extract the four "while True -> get_next_page() -> break" paging behaviors into a single method/function.